File: src/pages/leafs/match.py
"""Part of KgF.

MIT License
Copyright (c) 2017 LordKorea

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
of the Software, and to permit persons to whom the Software is furnished to do
so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

from kgf import kconfig
from model.match import Match
from model.participant import Participant
from pages.controller import Controller
from pages.templates.engine import Parser


class MatchController(Controller):
    """Handles the /match leaf."""

    def __init__(self):
        """Constructor."""
        super().__init__()

        # Whether the logout link will be shown
        self._logout_shown = kconfig().get("login-required", True)

        # Only allow logged-in users in a match
        self.add_access_restriction(self.check_access)
        self.set_permission_fail_handler(self.fail_permission)

        # Match creation endpoint
        self.add_endpoint(self.create_match,
                          path_restrict={"create"},
                          params_restrict={"deckupload"})
        self.add_endpoint(self.fail_permission, path_restrict={"create"})

        # Match joining endpoint
        self.add_endpoint(self.join_match, path_restrict={"join"})

        # Match leaving
        self.add_endpoint(self.abandon_match, path_restrict={"abandon"})

        self.add_endpoint(self.match)

    def check_access(self, session, path, params, headers):
        """Checks whether the user is logged in (and in a match).

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            bool: True if the user is logged in and in a match (if required).
        """
        if "login" not in session:
            return False

        # Exceptions to the 'user has to be in a match rule':
        #  - Joining a match
        #  - Creating a match
        is_exempt = ("create" in path) or ("join" in path)
        if not is_exempt and Match.get_match_of_player(session["id"]) is None:
            return False

        return True

    def fail_permission(self, session, path, params, headers):
        """Handles unauthorized clients.

        Redirects the client to the dashboard.

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            tuple: Returns 1) the HTTP status code 2) the HTTP headers to be
                sent and 3) the response to be sent to the client.
        """
        return (303,  # 303 See Other
                {"Location": "/dashboard"},
                "")

    def abandon_match(self, session, path, params, headers):
        """Handles the request to abandon a match.

        Redirects the client to the dashboard.

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            tuple: Returns 1) the HTTP status code 2) the HTTP headers to be
                sent and 3) the response to be sent to the client.
        """
        # Leave the match
        match = Match.get_match_of_player(session["id"])
        match.abandon_participant(session["id"])

        return (303,  # 303 See Other
                {"Location": "/dashboard"},
                "")

    def join_match(self, session, path, params, headers):
        """Handles the request to join an existing match.

        Redirects the client to the match view.

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            tuple: Returns 1) the HTTP status code 2) the HTTP headers to be
                sent and 3) the response to be sent to the client.
        """
        if Match.get_match_of_player(session["id"]) is not None:
            # The user is already in a match
            return (303,  # 303 See Other
                    {"Location": "/match"},
                    "")

        # Get the transmitted match ID
        try:
            id = int(path[-1])
        except ValueError:
            return self.fail_permission(session, path, params, headers)

        # Get the match from the pool
        match = Match.get_by_id(id)
        if match is None or match.has_started():
            return self.fail_permission(session, path, params, headers)

        # Put the player into the match
        part = Participant(session["id"], session["nickname"])
        match.add_participant(part)

        return (303,  # 303 See Other
                {"Location": "/match"},
                "")

    def create_match(self, session, path, params, headers):
        """Handles the request to create a match.

        Redirects either to the match view (on success) or to the dashboard
        (when the deck is too big).

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            tuple: Returns 1) the HTTP status code 2) the HTTP headers to be
                sent and 3) the response to be sent to the client.
        """
        if Match.get_match_of_player(session["id"]) is not None:
            # The user already is in a match
            return (303,  # 303 See Other
                    {"Location": "/match"},
                    "")

        if not params["deckupload"].filename:
            # Make sure the uploaded thing is a file
            return self.fail_permission(session, path, params, headers)

        # Check the size of the upload
        fp = params["deckupload"].file
        fp.seek(0, 2)
        size = fp.tell()
        fp.seek(0)
        if size > 200 * 1024:
            return (303,  # 303 See Other
                    {"Location": "/dashboard/filesizefail"},
                    "")

        # Create a new match
        match = Match()

        # Create the deck from the upload
        match.create_deck(fp.read().decode())

        # Add the participant to the match
        part = Participant(session["id"], session["nickname"])
        match.add_participant(part)

        # Make the match available for others
        match.put_in_pool()

        return (303,  # 303 See Other
                {"Location": "/match"},
                "")

    def match(self, session, path, params, headers):
        """Handles a request of the match view.

        Serves the match template.

        Args:
            session (obj): The session data of the client.
            path (list): The path of the request.
            params (dict): The HTTP POST parameters.
            headers (dict): The HTTP headers that were sent by the client.

        Returns:
            tuple: Returns 1) the HTTP status code 2) the HTTP headers to be
                sent and 3) the response to be sent to the client.
        """
        # Populate symbol table
        symtab = {"theme": session["theme"],
                  "showLogout": True if self._logout_shown else None}

        # Parse the template
        data = Parser.get_template("./res/tpl/match.html", symtab)
        return (200,  # 200 OK
                {"Content-Type": "text/html; charset=utf-8"},
                data)

File: src/res/js/match/match.js
/**
 * Part of KgF.
 *
 * MIT License
 * Copyright (c) 2017-2018 LordKorea
 * Copyright (c) 2018 Arc676/Alessandro Vinciguerra
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */
"use strict";

(function(){
  let countDown = 0
  let ending = false
  let numGaps = 0
  let allowChoose = false
  let allowPick = false
  let isSpectator = false
  let isPicker = false
  let externalUpdateAllowed = true
  let handResolver = new Map([
    ["OBJECT", new Map()],
    ["VERB", new Map()]
  ])
  let numSelected = 0
  let selectedCards = new Map()

  /**
   * Loads the match's status.
   */
  function loadStatus() {
    $.ajax({
      method: "GET",
      url: "/api/status",
      dataType: "json",
      success: updateStatus,
      error: (x, e, f) => console.log(`/api/status error: ${e} ${f}`)
    })
  }

  /**
   * Updates the status of the match.
   *
   * @param data The JSON data which was retrieved from the API.
   */
  function updateStatus(data) {
    countDown = data.timer
    $("#matchstatus").html(data.status)
    ending = data.ending
    numGaps = data.gaps
    allowChoose = data.allowChoose
    allowPick = data.allowPick
    isSpectator = data.isSpectator

    $("#cardcount").html("Selected: " + numSelected + " of " + numGaps);
    $("#cardcount").toggleClass("invisible", !allowChoose)

    if (isSpectator) {
      $(".match-hand").addClass("invisible")
    }
    isPicker = data.isPicker

    updateMatchStatement(data.hasCard, data.cardText || "Waiting...")

    updateCountdown()
    if (ending && countDown < 3) {
      window.location.assign("/dashboard")
    }
  }

  /**
   * Updates the countdown timer.
   */
  function updateCountdown() {
    let minutes = (Math.max(0, Math.floor(countDown / 60)) + "")
    minutes = minutes.padStart(2, "0")
    let seconds = (Math.max(0, countDown) % 60 + "").padStart(2, "0")
    $("#countdown").text(`${minutes}:${seconds}`)
  }

  /**
   * Updates the match's statement card.
   *
   * @param hasCard Whether the match has a statement card.
   * @param text The text of the statement card.
   */
  function updateMatchStatement(hasCard, text) {
    let elem = $("#match-statement")
    elem.html(applyCardTextFormat(text))
    if (hasCard) {
      elem.addClass("statement-card").removeClass("system-card")
    } else {
      elem.addClass("system-card").removeClass("statement-card")
    }
  }

  /**
   * Loads the match's cards (hand and played).
   */
  function loadCards() {
    $.ajax({
      method: "GET",
      url: "/api/cards",
      dataType: "json",
      success: updateCards,
      error: (x, e, f) => console.log(`/api/cards error: ${e} ${f}`)
    })
  }

  /**
   * Updates the cards (hand and played) of the match.
   *
   * @param data The JSON data which was retrieved from the API.
   */
  function updateCards(data) {
    let sentHand = new Map([
      ["OBJECT", new Map()],
      ["VERB", new Map()]
    ])
    if (data.hasOwnProperty("hand")) {
      for (let type of sentHand.keys()) {
        for (let key in data.hand[type]) {
          sentHand.get(type).set(key, data.hand[type][key])
        }
        updateHand(sentHand.get(type), type)
      }
    }
    updatePlayed(data.played)
  }

  /**
   * Updates the played card area.
   *
   * @param data The played cards data.
   */
  function updatePlayed(data) {
    let area = $(".match-played-cards")
    let sets = area.children(".match-played-set")
    if (sets.length === 0) {
      area.prepend(createCardSet(1))
      sets = area.children(".match-played-set")
    }
    let key = 0
    for (let dataSet of data) {
      while (key > sets.length) {
        sets.last().after(createCardSet(key))
        sets = area.children(".match-played-set")
      }

      let set = sets.eq(key - 1)
      set.empty()
      set.css("display", "none")
      for (let card of dataSet) {
        let elem = createPlayedCard(card)
        set.append(elem)
        set.css("display", "flex")
      }

      key++
    }
  }

  /**
   * Creates a played card element.
   *
   * @param card The card data for the card that should be created.
   */
  function createPlayedCard(card) {
    if (card.hasOwnProperty("redacted")) {
      return (
        $("<div></div>", {"class": "card-base system-card"}).prepend(
          $("<div></div>", {"class": "card-label"}).text("?")
        )
      )
    } else {
      return (
        $("<div></div>", {
          "class": `card-base ${card.type.toLowerCase()}-card`
        }).html(applyCardTextFormat(card.text))
      )
    }
  }

  /**
   * Creates a card set for the participant with the given ID.
   *
   * @param key The ID of the participant.
   * @return The element for the set.
   */
  function createCardSet(key) {
    $(document).on("click", `#played-id-${key}`, {}, () => clickPlayed(key))
    return (
      $("<div></div>", {
        "class": "match-played-set",
        "id": `played-id-${key}`
      }).css("display", "none")
    )
  }

  /**
   * Handles clicking on a card set.
   *
   * @param key The order key of the card set.
   */
  function clickPlayed(key) {
    if (!allowPick) {
      return
    }
    $.ajax({
      method: "POST",
      url: "/api/pick",
      data: {"playedId": key}
    })
  }

  /**
   * Updates the hand of the client.
   *
   * @param hand The hand that was sent by the server.
   * @param type The card type of the hand.
   */
  function updateHand(hand, type) {
    let lowerType = type.toLowerCase()
    let container = $(`#${lowerType}-hand`)
    let resolver = handResolver.get(type)

    // Unselect hand cards that have been unselected on the server side.
    // This is not possible one second after a newly chosen card is marked.
    if (externalUpdateAllowed) {
      for (let [id, data] of hand.entries()) {
        if (data.chosen === null && resolver.has(id)) {
          unselectHandCard(resolver.get(id))
        }
      }
    }

    // Add new cards
    for (let id of hand.keys().filter(x => !resolver.has(x))) {
      let elem = createHandCard(lowerType, id, hand.get(id).text)
      if (hand.get(id).chosen !== null) {
        selectHandCard(elem, hand.get(id).chosen)
      }
      container.append(elem)
      resolver.set(id, elem)
    }

    // Delete old cards (uses array because of concurrent modification)
    for (let id of Array.from(resolver.keys().filter(x => !hand.has(x)))) {
      unselectHandCard(resolver.get(id))
      resolver.get(id).remove()
      resolver.delete(id)
    }

    if (resolver.size > 0) {
      container.children(".system-card").remove()
    } else {
      container.empty().append(createSystemCard())
    }

    $(".hand-set").find(".card-base").toggleClass("grey-card", isPicker)
  }

  /**
   * Marks a card with a selection watermark.
   *
   * @param card The card that should be watermarked.
   * @param k The number that should be put on the card.
   */
  function selectHandCard(card, k) {
    if (card.children(".card-label").length === 0) {
      card.prepend($("<div></div>").addClass("card-label").text("" + (k + 1)))
    } else {
      card.children(".card-label").text("" + (k + 1))
    }
    card.addClass("card-selected")
    selectedCards.set(k, card)
    numSelected++;
  }

  /**
   * Removes the watermark from the given card.
   *
   * @param card The card that should have its watermark removed.
   */
  function unselectHandCard(card) {
    card.children(".card-label").remove()
    card.removeClass("card-selected")
    for (let [id, entry] of selectedCards.entries()) {
      if (entry === card) {
        selectedCards.delete(id)
        numSelected--
        break
      }
    }
  }

  /**
   * Creates a system card.
   *
   * @return The system card element.
   */
  function createSystemCard() {
    return (
      $("<div></div>", {
        "class": "card-base system-card"
      }).text("No cards on your hand.")
    )
  }

  /**
   * Creates a hand card element.
   *
   * @param type The card type.
   * @param id The ID of the hand card.
   * @param text The text of the hand card.
   * @return The hand card element.
   */
  function createHandCard(type, id, text) {
    $(document).on("click", `#hand-id-${id}`, {}, () => toggleSelect(id))
    return (
      $("<div></div>", {
        "class": `card-base ${type}-card`,
        "id": `hand-id-${id}`
      }).html(applyCardTextFormat(text))
    )
  }

  /**
   * Toggles selection of the hand card with the given ID.
   *
   * @param id The ID of the hand card.
   */
  function toggleSelect(id) {
    if (!allowChoose) {
      return
    }
    $.ajax({
      method: "POST",
      url: "/api/choose",
      data: {"handId": id},
      success: () => updateSelection(id),
      error: (x, e, f) => console.log(`/api/choose error: ${e} ${f}`)
    })
  }

  /**
   * Updates the hand selection.
   *
   * @param id The ID of the card.
   */
  function updateSelection(id) {
    externalUpdateAllowed = false
    setTimeout(() => externalUpdateAllowed = true, 1000)

    let remove = false
    let card = handResolver.get("OBJECT").get(id)
    card = card || handResolver.get("VERB").get(id)

    // Remove card choices
    for (let i = 0; i < 4; i++) {
      if (!selectedCards.has(i)) {
        continue
      }
      if (selectedCards.get(i) === card || remove) {
        unselectHandCard(selectedCards.get(i))
        remove = true
      }
    }

    // Add a new card
    if (!remove) {
      if (numSelected < numGaps) {
        selectHandCard(card, numSelected)
      }
    }
  }

  /**
   * Picks and shows the hand tab with the given ID.
   *
   * @param tab The ID of the tab.
   */
  function pickTab(tab) {
    $(".hand-tab-header").removeClass("active-tab-header")
    $(".hand-set").css("display", "none")
    $(`.${tab}`).css("display", "inherit")
    $(`#${tab}`).addClass("active-tab-header")
  }

  /**
   * Chooses the actions tab.
   */
  function chooseActionsTab() {
    pickTab("tab-actions")
  }

  /**
   * Chooses the objects tab.
   */
  function chooseObjectsTab() {
    pickTab("tab-objects")
  }

  /**
   * Sends POST request to skip the remaining time
   */
  function skipTime() {
    $.ajax({
      method: "POST",
      url: "/api/skip"
    })
  }

  /**
   * Toggle hand visibility
   */
   function toggleHand() {
     $("#hand-container").hide()
   }

  setInterval(loadStatus, 1000)
  loadStatus()
  setInterval(loadCards, 1000)
  loadCards()
  pickTab("tab-actions")
  $("#tab-actions").click(chooseActionsTab)
  $("#tab-objects").click(chooseObjectsTab)
  $("#skip-button").click(skipTime)
  $("#toggle-hand-button").click(toggleHand)
})()
